cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,506 @@
+		klog.Error(err.Error())
+		return nil, err
+	}
+
+	var candidate podCandidate
+	candidates, err := rm.findAllocationPodCandidates(pendingPods)
+	if err != nil {
+		klog.Error(err.Error())
+		return nil, err
+	}
+
+	numCandidates := len(candidates)
+	switch numCandidates {
+	case 0:
+		// fine, this typically happens when deployment is deleted before PODs start
+		klog.V(4).Info("zero pending pods")
+		return nil, errZeroPending
+	case 1:
+		// perfect, only one option
+		klog.V(4).Info("only one pending pod")
+		candidate = candidates[0]
+		if _, ok := candidate.pod.Annotations[gasCardAnnotation]; !ok {
+			klog.Warningf("Pending POD annotations from scheduler not yet visible for pod %v", candidate.pod.Name)
+			return nil, errRetry
+		}
+	default: // > 1 candidates, not good, need to pick the best
+		// look for scheduler timestamps and sort by them
+		klog.V(4).Infof("%v pods pending, picking oldest", numCandidates)
+		timestampedCandidates := []podCandidate{}
+		for _, candidate := range candidates {
+			if _, ok := pendingPods[candidate.name].Annotations[gasTSAnnotation]; ok {
+				timestampedCandidates = append(timestampedCandidates, candidate)
+			}
+		}
+		sort.Slice(timestampedCandidates,
+			func(i, j int) bool {
+				return pendingPods[timestampedCandidates[i].name].Annotations[gasTSAnnotation] <
+					pendingPods[timestampedCandidates[j].name].Annotations[gasTSAnnotation]
+			})
+		if len(timestampedCandidates) == 0 {
+			klog.Warning("Pending POD annotations from scheduler not yet visible")
+			return nil, errRetry
+		}
+		candidate = timestampedCandidates[0]
+	}
+
+	candidate.pod = pendingPods[candidate.name]
+	return &candidate, nil
+}
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list
+
+// getNodePendingGPUPods returns a map of pod names -> pods that are pending and use the gpu.
+func (rm *resourceManager) getNodePendingGPUPods() (map[string]*v1.Pod, error) {
+	selector, err := fields.ParseSelector("spec.nodeName=" + rm.nodeName +
+		",status.phase=" + string(v1.PodPending))
+
+	if err != nil {
+		klog.Error(err.Error())
+		return nil, err
+	}
+
+	pendingPodList, err := rm.clientset.CoreV1().Pods(v1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
+		FieldSelector: selector.String(),
+	})
+
+	if err != nil {
+		klog.Error(err.Error())
+		return nil, err
+	}
+
+	// make a map ouf of the list, accept only GPU-using pods
+	pendingPods := make(map[string]*v1.Pod)
+	for i := range pendingPodList.Items {
+		pod := &pendingPodList.Items[i]
+
+		if numGPUUsingContainers(pod) > 0 {
+			pendingPods[pod.Name] = pod
+		}
+	}
+
+	return pendingPods, nil
+}
+
+// findAllocationPodCandidates returns a slice of all potential allocation candidate pods.
+// This goes through the PODs listed in the podresources grpc service and finds those among pending
+// pods which don't have all GPU devices allocated.
+func (rm *resourceManager) findAllocationPodCandidates(pendingPods map[string]*v1.Pod) (candidates []podCandidate, err error) {
```

Comment:
This function doesn't touch `rm`'s state and as such can be a free function.

Reply:
indeed, done